dbt/adapters/redshift/impl.py

```diff
@@ -183,3 +186,12 @@ def generate_python_submission_response(self, submission_result: Any) -> Adapter
     def debug_query(self):
         """Override for DebugTask method"""
         self.execute("select 1 as id")
+
+    def calculate_freshness_from_metadata_batch(
+        self,
+        sources: List[BaseRelation],
+        macro_resolver: Optional[MacroResolverProtocol] = None,
+    ) -> Tuple[List[Optional[AdapterResponse]], Dict[BaseRelation, FreshnessResponse]]:
+        conn = self.connections.get_if_exists()
+        self.connections.open(conn)
```

Comment:
I don't necessarily think this is the right implementation approach, but the issue I was running into reusing the `calculate_freshness_from_metadata_batch` from the `BaseAdapter` was that `dbt-core` is expecting to be able to call this method/macro prior to the per-node (source node in this case) job queue execution in the `source freshness` task. Presumably the redshift connection had not yet been opened, which leads to the following error without these changes ([logs](https://github.com/dbt-labs/dbt-redshift/actions/runs/8475089564/job/23222602883)): 

```python
dbt_common.exceptions.base.DbtRuntimeError: Runtime Error
E             connection is closed
```

Is there a more idiomatic way to ensure a connection is open prior to execution a macro within an adapter, or for dbt-redshift specifically? cc @colin-rogers-dbt @mikealfare

Reply:
there certainly _should_ be a more idiomatic approach

Reply:
Resolved offline / in dbt-adapters PR: https://github.com/dbt-labs/dbt-adapters/pull/127/files#diff-e2306b841793e5b07177010e8920c59163df1536306fc76e7f6a09d36dcae4d3R1078-R1081

Ended up introducing a `needs_conn` parameter to `execute_macro` and setting it to True in the base adapter implementation to avoid the override in dbt-redshift